### PR TITLE
Put a size limit on the resource cache.

### LINKED ISSUE #78: Make ResourceCache scalable
IIUC, webpkgserver always runs with an [in-memory resource cache](https://github.com/google/webpackager/blob/76848efd85256189c4f2ebb4c4f977171b011d1d/config.go#L124), which is [unbounded and lacking expiration or eviction](https://github.com/google/webpackager/blob/76848efd85256189c4f2ebb4c4f977171b011d1d/resource/cache/memory.go#L37-L40). (Expired entries are never deleted from memory unless they are replaced.) This cache persists for the [life of the server](https://github.com/google/webpackager/blob/76848efd85256189c4f2ebb4c4f977171b011d1d/server/handler.go#L60).

This is OK only for a site with a small number of resources. We need to add some configuration parameters to make this work for larger sites. Some rough ideas:

 - Max size of the cache (including `0` to disable). For starters, we can do it by # of entries, but # of bytes would be a nice addition.
 - Option to use the [file-based cache](https://github.com/google/webpackager/tree/76848efd85256189c4f2ebb4c4f977171b011d1d/resource/cache/filewrite) as a backend to the memory cache, so it can be shared between replicas. This would reduce # of fetches.

### PR DESCRIPTION
**Pull request by twifkak**
Change the default resource cache for webpkgserver to be size-bound (200
entries as a reasonable-seeming default), with options to make it unbounded or
to disable it.

This required changing Packager.RunForRequest to return the Resource it
creates, rather than having the caller rely on it being present as a side
effect in the ResourceCache. Otherwise, a cache eviction policy would be
incompatible with parallel calls to server.Handler.handleDocImpl (short of a
rather large critical section that would make it mostly not parallel).

Fixes #78.

### CODE REVIEW

File: resource/cache/bounded.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+	"github.com/google/webpackager/resource"
+
+	lru "github.com/hashicorp/golang-lru"
+	"golang.org/x/xerrors"
+)
+
+// NewBoundedCache returns a new ResourceCache that stores Resources in memory,
+// with an eviction policy after `size` entries. size must be positive.
+func NewBoundedCache(size int) ResourceCache {
```
Comment (banaag) [inline resource/cache/bounded.go:29]: nit: Should it be NewBoundedInMemoryCache? I don't have a strong preference.

File: resource/cache/bounded.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	c.lru.Add(key, value)
+}
+
+func (c lruCache) Get(key interface{}) (value interface{}, ok bool) {
+	return c.lru.Get(key)
+}
+
+type boundedCache struct {
+	cache
+}
```
Comment (banaag) [inline resource/cache/bounded.go:76]: You might want to add an interface check here: var _ cache = (*boundedCache)(nil)
See: https://stackoverflow.com/questions/17994519/golang-interface-compliance-compile-type-check

File: resource/cache/bounded.go (hunk 1 above)
Comment (twifkak) [reply resource/cache/bounded.go]: Sure, done.

File: resource/cache/bounded.go (hunk 2 above)
Comment (twifkak) [reply resource/cache/bounded.go:56]: Heh, clever. Done.

**Review by banaag — APPROVED** [review]

**Comment by twifkak** [comment]
Thanks! Merging.

**Pull request merged**